yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala

```diff
@@ -82,6 +93,102 @@ class YarnSparkHadoopUtil extends SparkHadoopUtil {
     if (credentials != null) credentials.getSecretKey(new Text(key)) else null
   }
 
+  override def setPrincipalAndKeytabForLogin(principal: String, keytab: String): Unit = {
+    loginPrincipal = Option(principal)
+    keytabFile = Option(keytab)
+  }
+
+  private[spark] override def scheduleLoginFromKeytab(
+    callback: (SerializableBuffer)  => Unit): Unit = {
+
+    loginPrincipal match {
+      case Some(principal) =>
+        val keytab = keytabFile.get
+        val remoteFs = FileSystem.get(conf)
+        val remoteKeytabPath = new Path(
+          remoteFs.getHomeDirectory, System.getenv("SPARK_STAGING_DIR") + Path.SEPARATOR + keytab)
+        val localFS = FileSystem.getLocal(conf)
+        // At this point, SparkEnv is likely no initialized, so create a dir, put the keytab there.
+        val tempDir = Utils.createTempDir()
+        val localURI = new URI(tempDir.getAbsolutePath + Path.SEPARATOR + keytab)
+        val qualifiedURI = new  URI(localFS.makeQualified(new Path(localURI)).toString)
+        FileUtil.copy(
```

Comment:
Frankly, the only reason I am doing this manually is because I could really not understand how the stuff that is added for localization can be accessed from the AM. I really could not figure out the code that copies stuff to the cache and how they can be accessed from the AM. (The YARN docs on this seem really lacking, and I could not find such an example in the AM code either). The current "downloading" code seem to be all in the Executor, so getting the AM to download stuff seemed tricky too. If you have pointers, I will gladly change this.

Also, I think this can really be changed later as well, since this is more of an implementation detail.